lib/services/role.go

```diff
@@ -2848,7 +2849,43 @@ type checkAccessParams struct {
 	silent                bool
 }
 
-func (set RoleSet) checkAccessToRuleImpl(p checkAccessParams) error {
+type accessExplicitlyDenied struct {
+	inner error
+}
+
+// AccessExplicitlyDenied is an error type that indicates an AccessDenied error
+// where a deny rule matched and access is explicitly denied, in contrast to
+// cases where there is no matching deny or allow rule and access is only
+// implicitly denied.
+func AccessExplicitlyDenied(inner error) error {
+	return &accessExplicitlyDenied{inner}
+}
+
+// IsAccessExplicitlyDenied returns true if any of the errors in err's chain is
+// an AccessExplicitlyDenied error.
+func IsAccessExplicitlyDenied(err error) bool {
+	var target *accessExplicitlyDenied
+	return errors.As(err, &target)
```

Comment:
If we only need a book and don't care about the actual target error, can we use errors.Is instead?

Reply:
I think that would get the semantics wrong. I want to match any AccessExplicitlyDenied error regardless of the underlying/`inner` error, but `errors.Is` tests for equality with a specific value. Of course, I could make it work by implementing an `Is` method, but I think having to do that kind of shows that you're doing something non-standard.